client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1607 @@
+		let _ = self.to_controller.unbounded_send(Action::DisconnectPeer(peer_id));
+	}
+
+	/// Notify about incoming connection. [`ProtocolController`] will either accept or reject it.
+	pub fn incoming_connection(&self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		let _ = self
+			.to_controller
+			.unbounded_send(Action::IncomingConnection(peer_id, incoming_index));
+	}
+
+	/// Notify that connection was dropped (either refused or disconnected).
+	pub fn dropped(&self, peer_id: PeerId) {
+		let _ = self.to_controller.unbounded_send(Action::Dropped(peer_id));
+	}
+
+	/// Get the list of reserved peers.
+	pub fn reserved_peers(&self, pending_response: oneshot::Sender<Vec<PeerId>>) {
+		let _ = self.to_controller.unbounded_send(Action::ReservedPeers { pending_response });
+	}
+}
+
+/// Direction of a connection
+#[derive(Clone, Copy, Debug)]
+enum Direction {
+	Inbound,
+	Outbound,
+}
+
+/// Status of a connection with a peer.
+#[derive(Clone, Debug)]
+enum PeerState {
+	/// We are connected to the peer.
+	Connected(Direction),
+	/// We are not connected.
+	NotConnected,
+}
+
+impl PeerState {
+	/// Returns true if we are connected with the node.
+	fn is_connected(&self) -> bool {
+		matches!(self, PeerState::Connected(_))
+	}
+}
+
+impl Default for PeerState {
+	fn default() -> PeerState {
+		PeerState::NotConnected
+	}
+}
+
+/// Side of [`ProtocolHandle`] responsible for all the logic. Currently all instances are
+/// owned by [`crate::Peerset`], but they should eventually be moved to corresponding protocols.
+#[derive(Debug)]
+pub struct ProtocolController<PeerStoreHandle> {
+	/// Set id to use when sending connect/drop requests to `Notifications`.
+	// Will likely be replaced by `ProtocolName` in the future.
+	set_id: SetId,
+	/// Receiver for messages from [`ProtocolHandle`].
+	from_handle: TracingUnboundedReceiver<Action>,
+	/// Number of occupied slots for incoming connections (not counting reserved nodes).
+	num_in: u32,
+	/// Number of occupied slots for outgoing connections (not counting reserved nodes).
+	num_out: u32,
+	/// Maximum number of slots for incoming connections (not counting reserved nodes).
+	max_in: u32,
+	/// Maximum number of slots for outgoing connections (not counting reserved nodes).
+	max_out: u32,
+	/// Connected regular nodes.
+	nodes: HashMap<PeerId, Direction>,
+	/// Reserved nodes. Should be always connected and do not occupy peer slots.
+	reserved_nodes: HashMap<PeerId, PeerState>,
+	/// Connect only to reserved nodes.
+	reserved_only: bool,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+	/// Peerset handle for checking peer reputation values and getting connection candidates
+	/// with highest reputation.
+	peer_store: PeerStoreHandle,
+}
+
+impl<PeerStoreHandle: PeerReputationProvider> ProtocolController<PeerStoreHandle> {
```

Comment:
This could also take `Box<dyn PeerReputationProvider>` so you wouldn't have to make it generic over `PeerStoreHandle` but if you prefer it this way, sall good.